modMcf/src/org/aion/mcf/vm/types/DataWord.java

```diff
@@ -54,7 +54,31 @@ public DataWord(byte[] data) {
     }
 
     public DataWord(BigInteger num) {
-        this(num.toByteArray());
+        // NOTE: DataWord.value() produces a signed positive BigInteger. The byte array
+        // representation of such a number must prepend a zero byte so that this can be decoded
+        // correctly. This means that a 16-byte array with a non-zero starting bit will become 17
+        // bytes when BigInteger::toByteArray is called, and therefore we must remove any leading
+        // zero bytes from this representation for full compatibility.
+        this(removeLargeBigIntegerLeadingZeroByte(num));
+    }
+
+    /**
+     * Similar to {@code stripLeadingZeroes} but more specialized to be more efficient in a specific
+     * necessary situation.
+     *
+     * <p>Essentially this method will always return {@code number.toByteArray()} UNLESS this byte
+     * array is length {@value BYTES} + 1 (that is, 17), and its initial byte is a zero byte. In
+     * this single case, the leading zero byte will be stripped.
+     *
+     * @param number The {@link BigInteger} whose byte array representation is to be possibly
+     *     truncated.
+     * @return The re-formatted {@link BigInteger#toByteArray()} representation as here specified.
+     */
+    private static byte[] removeLargeBigIntegerLeadingZeroByte(BigInteger number) {
+        byte[] bytes = number.toByteArray();
+        return ((bytes.length == (DataWord.BYTES + 1)) && (bytes[0] == 0x0))
+                ? Arrays.copyOfRange(bytes, 1, bytes.length)
+                : bytes;
```

Comment:
how to handle the case when the Bignumber input more than 16 bytes?

Reply:
This byte array gets passed into the constructor, which performs the check